### PR TITLE
Added ResizeBilinear op for tf

### PR DESCRIPTION
**Pull request by dmonterom**
### This pullrequest changes

Added ResizeBilinear op for tensorflow importer.

### CODE REVIEW

File: modules/dnn/include/opencv2/dnn/all_layers.hpp [hunk 1]
```diff
@@ -575,6 +575,12 @@ CV__DNN_EXPERIMENTAL_NS_BEGIN
         static Ptr<ResizeNearestNeighborLayer> create(const LayerParams& params);
     };
 
+    class CV_EXPORTS ResizeBilinearLayer : public Layer
```
Comment (dkurt) [inline modules/dnn/include/opencv2/dnn/all_layers.hpp]: @dmonterom, what do you think about renaming `ResizeNearestNeighborLayer` to `ResizeLayer` and use interpolation strategy as an additional flag? Please add a test similar to nearest neighbor layer

https://github.com/opencv/opencv/blob/5b868ccd829975da5372bf330994553e176aee09/modules/dnn/test/test_tf_importer.cpp#L325-L328

https://github.com/opencv/opencv_extra/blob/b467369489529d8eb2e847c13b117ec4e253a638/testdata/dnn/tensorflow/generate_tf_models.py#L258-L262

Comment (dmonterom) [reply modules/dnn/include/opencv2/dnn/all_layers.hpp]: Hi @dkurt, thats a good idea, I will change it as soon as possible. I will also add the new test soon.

Comment (dkurt) [reply modules/dnn/include/opencv2/dnn/all_layers.hpp]: @dmonterom, thank you! It also can help us to resolve an issue https://github.com/opencv/opencv/issues/10985 with `Interp` layer from some of Caffe's forks smoother.

File: modules/dnn/src/tensorflow/tf_importer.cpp [hunk 2]
```diff
@@ -1469,11 +1469,29 @@ void TFImporter::populateNet(Net dstNet)
 
             layerParams.set("height", outSize.at<int>(0, 0));
             layerParams.set("width", outSize.at<int>(0, 1));
+            layerParams.set("interpolation", 0);
```
Comment (dkurt) [inline modules/dnn/src/tensorflow/tf_importer.cpp]: Enumerations `INTER_NEAREST` and `INTER_LINEAR` are also integers. You can just set them as arguments, `layerParams.set("interpolation", INTER_NEAREST);` and use `interpolation` like `resize(getPlane(inp, n, ch), getPlane(out, n, ch), Size(outWidth, outHeight), 0, 0, interpolation);`

File: modules/dnn/src/tensorflow/tf_importer.cpp [hunk 3]
```diff
@@ -1469,11 +1469,29 @@ void TFImporter::populateNet(Net dstNet)
...
             if (hasLayerAttr(layer, "align_corners"))
                 layerParams.set("align_corners", getLayerAttr(layer, "align_corners").b());
 
-            int id = dstNet.addLayer(name, "ResizeNearestNeighbor", layerParams);
+            int id = dstNet.addLayer(name, "Resize", layerParams);
+            layer_id[name] = id;
+
+            connect(layer_id, dstNet, parsePin(layer.input(0)), id, 0);
+        }
+        else if (type == "ResizeBilinear")
```
Comment (dkurt) [inline modules/dnn/src/tensorflow/tf_importer.cpp]: Body of that case is absolutely the same. Can you just extend the previous one by `else if (type == "ResizeNearestNeighbor" || type == "ResizeBilinear")` and vary "interpolation" flag?

File: modules/dnn/include/opencv2/dnn/all_layers.hpp (hunk 1 above)
Comment (dmonterom) [reply modules/dnn/include/opencv2/dnn/all_layers.hpp]: Hey @dkurt, I made the changes (I also tested the two layers and everything seems ok). If you agree with them i will replace the resize nearest neighbor test with a new one.

Comment (dkurt) [reply modules/dnn/include/opencv2/dnn/all_layers.hpp]: @dmonterom please add `*_net`, `*_in` and `*_out` files generated by script as a PR to opencv_extra with modified version of the script.

File: modules/dnn/src/init.cpp [hunk 4]
```diff
@@ -83,7 +83,7 @@ void initializeLayerFactory()
     CV_DNN_REGISTER_LAYER_CLASS(Concat,         ConcatLayer);
     CV_DNN_REGISTER_LAYER_CLASS(Reshape,        ReshapeLayer);
     CV_DNN_REGISTER_LAYER_CLASS(Flatten,        FlattenLayer);
-    CV_DNN_REGISTER_LAYER_CLASS(ResizeNearestNeighbor, ResizeNearestNeighborLayer);
+    CV_DNN_REGISTER_LAYER_CLASS(Resize, ResizeLayer);
```
Comment (dkurt) [inline modules/dnn/src/init.cpp]: Please align `ResizeLayer` with the other layers names.

File: modules/dnn/src/tensorflow/tf_importer.cpp [hunk 5]
```diff
@@ -1462,18 +1462,22 @@ void TFImporter::populateNet(Net dstNet)
...
-        else if (type == "ResizeNearestNeighbor")
+        else if (type == "ResizeNearestNeighbor" || type == "ResizeBilinear")
         {
             Mat outSize = getTensorContent(getConstBlob(layer, value_id, 1));
             CV_Assert(outSize.type() == CV_32SC1, outSize.total() == 2);
 
             layerParams.set("height", outSize.at<int>(0, 0));
             layerParams.set("width", outSize.at<int>(0, 1));
+            if (type == "ResizeNearestNeighbor")
+                layerParams.set("interpolation", 0);
```
Comment (dkurt) [inline modules/dnn/src/tensorflow/tf_importer.cpp]: I think `INTER_NEAREST` or `INTER_LINEAR` should be here.

Comment (dmonterom) [reply modules/dnn/src/tensorflow/tf_importer.cpp]: Hi, do you mean replacing the values by them? Then I should define them at the beginning or include imgproc.hpp, right?

Comment (dkurt) [reply modules/dnn/src/tensorflow/tf_importer.cpp]: Why? They are just defined in OpenCV: https://docs.opencv.org/master/da/d54/group__imgproc__transform.html#ga5bb5a1fea74ea38e1a5445ca803ff121 . And please do not forget to add a test.

**Comment by dmonterom** [comment]
Hi @dkurt, I have created a test for resize bilinear using the generator script, but when i run the test it fails:
 ```
error: Expected: (normL1) <= (l1), actual: 0.122411 vs 1e-05
error: Expected: (normInf) <= (lInf), actual: 1.06526 vs 0.0001
```
I have also regenerate the resize nearest one but it successes, so its not about my tensorflow version.
Maybe they are using different bilinear interpolation approaches?

**Comment by dkurt** [comment]
@dmonterom, that is why we need to test it. Perhaps the problem is in borders.Try to figure out what values are different. Test on small inputs like `1x5x5x1` and compare outputs.

**Comment by dmonterom** [comment]
@dkurt, I have been doing several tests with different sizes and they always fail. I found this information:
 [https://hackernoon.com/how-tensorflows-tf-image-resize-stole-60-days-of-my-life-aba5eb093f35](url)
[https://stackoverflow.com/questions/48628736/bilinear-interpolation-implentations-in-tensorflow-and-opencv?rq=1](url)
So it seems that there is a bug in the tensorflow resize op. Anyway, the actual implementation of the layer does the trick: I have imported blitznet with dnn (which has several resize_bilinear layers) and the results are very close to the ones I get with tensorflow.

**Comment by dkurt** [comment]
@dmonterom, it'd nice if you can attach the results of your experiments. We need to know if difference is only at boundaries or in corners or in a whole image. Thanks!

**Comment by dmonterom** [comment]
@dkurt, here is an example of blitznet with tensorflow and with OpenCV
harry-meghan-15
Here is the result with tensorflow:
tensorflow_result
And here with OpenCV:
opencv_result
As you can see, there is a small accuracy drop with OpenCV but the results are quite good.

**Comment by dkurt** [comment]
Thank you but I meant numerical experiments with resize of 5x5 inputs. We need to compare numbers first.

**Comment by dkurt** [comment]
@dmonterom, thank you for your work and research! It seems that TensorFlow's resize_bilinear is a quite different from OpenCV's INTER_LINEAR. Let me introduce you an experimental mechanic of custom layers. All that you need is just use changes from PR https://github.com/opencv/opencv/pull/11129 (work in progress for now). You may find a tutorial there.

File: modules/dnn/src/layers/resize_layer.cpp [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+            const float widthScale = static_cast<float>(inpWidth) / (outWidth);
+            const int numPlanes = inp.size[0] * inp.size[1];
+            CV_Assert(inp.isContinuous(), out.isContinuous());
+
+            for (int y = 0; y < outHeight; ++y)
+            {
+                float input_y = y * heightScale;
+                int y0 = static_cast<int>(input_y);
+                const float* inpData_row0 = (float*)inp.data + y0 * inpWidth;
+                const float* inpData_row1 = (y0 + 1 < inpHeight) ? (inpData_row0 + inpWidth) : inpData_row0;
```
Comment (alalek) [inline modules/dnn/src/layers/resize_layer.cpp]: Use `.ptr<float>(y0)`
and `.ptr<float>(std::min(y0 + 1, inpHeight - 1))`

File: modules/dnn/src/layers/resize_layer.cpp [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+                for (int x = 0; x < outWidth; ++x)
+                {
+                    float input_x = x * widthScale;
+                    int x0 = static_cast<int>(input_x);
+                    int x1 = std::min(x0 + 1, inpWidth - 1);
+
+                    float* outData = (float*)out.data + y * outWidth + x;
```
Comment (alalek) [inline modules/dnn/src/layers/resize_layer.cpp]: `.ptr()`

File: modules/dnn/src/layers/resize_layer.cpp [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+                    const float* inpData_row0_c = inpData_row0;
+                    const float* inpData_row1_c = inpData_row1;
+                    for (int c = 0; c < numPlanes; ++c)
+                    {
+                        *outData = inpData_row0_c[x0] +
+                            (input_y - y0) * (inpData_row1_c[x0] - inpData_row0_c[x0]) +
+                            (input_x - x0) * (inpData_row0_c[x1] - inpData_row0_c[x0] +
+                            (input_y - y0) * (inpData_row1_c[x1] - inpData_row0_c[x1] - inpData_row1_c[x0] + inpData_row0_c[x0]));
```
Comment (alalek) [inline modules/dnn/src/layers/resize_layer.cpp]: Why OpenCV `cv::resize()` with `BORDER_REPLICATE` doesn't work here? (answer: `0.5` pixel center)

Anyway, if you want to reuse OpenCV's `INTER_LINEAR` constant then it should have **the same behavior** as defined by this constant.

File: modules/dnn/src/layers/resize_layer.cpp (hunk 6 above)
Comment (dkurt) [reply modules/dnn/src/layers/resize_layer.cpp]: Thanks for a review! Replaced it.

File: modules/dnn/src/layers/resize_layer.cpp (hunk 8 above)
Comment (dkurt) [reply modules/dnn/src/layers/resize_layer.cpp]: @alalek, There are different topics that TenorFlow's resize_bilinear doesn't equal to OpenCV's one.

Replaced constants to string values.

Comment (vpisarev) [reply modules/dnn/src/layers/resize_layer.cpp]: @alalek, if we extend cv::resize in the master branch and add the optional floating-point shift, will that make cv::resize suitable for implementation of this layer?

Comment (alalek) [reply modules/dnn/src/layers/resize_layer.cpp]: Two different formulas are usually used for resize operation:

1. `x * widthScale`
2. Something like `(x + 0.5) * widthScale - 0.5`

The second formula is used in OpenVX, OpenCV's `INTER_LINEAR` resize (and probably generic `INTER_AREA`).
The first is used in OpenCV `INTER_NEAREST` (not accurate, but it is fast) / `INTER_AREA` (fast path).

Here, we need to use the first formula with "linear" interpolation (to match original framework operation).

Comment (vpisarev) [reply modules/dnn/src/layers/resize_layer.cpp]: @alalek, it means that cv::warpAffine can do that?

**Pull request merged**